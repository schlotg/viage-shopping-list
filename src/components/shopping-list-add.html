<div style="background-image: url(${this.logo}); width: 129px; height: 128px; margin-left: auto; margin-right: auto; margin-bottom: 20px;"></div>
<h3 class="not-mobile" style="margin-left: 6px; color: green">${this.id ? 'Edit' : 'Add Item'}
</h3>
<div class="quantity-container">
  <label class="label">Quantity</label>
  <input class="input" type="number" attach="quantity"/>
</div>
<div class="name-container">
  <label class="label">Name</label>
  <input class="input" type="text" attach="name"/>
</div>
<div class="description-container">
  <label class="label">Description</label>
  <input class="input" type="text" attach="description"/>
</div>
<div>
  <button class="save-button button green" attach="save">Save</button>
  <button class="button blue" attach="back">Back</button>
<div>