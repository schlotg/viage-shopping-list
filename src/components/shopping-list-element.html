<div class="not-mobile list-item">
  <input attach="enabled" type="checkbox" ${this.purchased ? "checked": ""} />
  <span style="width: 30px; display: inline-block">${this.quantity}</span>
  <span style="width: 20%; display: inline-block">${this.name}</span>
  <span style="width: 40%; display: inline-block">${this.description}</span>
  <button attach="delete" style="padding: 1px; float: right">Delete</button>
  <button attach="edit" style="padding: 1px; float: right; margin-right: 10px;">Edit</button>
</div>
<div attach="m_edit" class="mobile list-item">
  <input attach="m_enabled" type="checkbox" ${this.purchased ? "checked": ""} />
  <span style="width: 30px; display: inline-block">${this.quantity}</span>
  <span style="width: 50%; display: inline-block">${this.name}</span>
  <button attach="m_delete" style="padding: 1px; float: right">X</button>
  <div style="display: block; margin-top:20px;">
    <label style="font-weight:800">Description: </label>
    <span style="width: 40%;" >${this.description}</span>
  </div>
</div>
