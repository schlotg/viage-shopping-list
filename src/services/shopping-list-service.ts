import { Service } from 'viage';

export interface Item {
  name: string;
  description: string;
  quantity: number;
  enabled: boolean;
  _id?: string; // added by the service
}

class ShoppingListServiceSingleton extends Service {

  private list: Item[] = [];
  private count = 1;

  private generateId() {
    this.count += 1;
    return this.count.toString(16);
  }

  private findIndex(id: string): number {
    let i = 0;
    const list = this.list;
    for (; i < list.length; ++i) {
      if (list[i]._id === id) { break; }
    }
    return i;
  }

  constructor() {
    super();
    if (window.localStorage) {
      const json = window.localStorage.getItem('ShoppingList');
      if (json) {
        const obj = JSON.parse(json) || {};
        this.count = obj.count || 0;
        const list = obj.list || [];
        list.forEach((e: Item) => this.addItem(e));
      }
    }
  }

  addItem(item: Item) {
    item._id = item._id || this.generateId();
    this.list.push(item);
    this.save();
  }

  removeItem(id: string) {
    const i = this.findIndex(id);
    this.list.splice(i, 1);
    this.save();
  }

  getItem(id: string): Item {
    const i = this.findIndex(id);
    return this.list[i];
  }

  forEach(cb: (item: Item, i?: number) => void) {
    this.list.forEach(cb);
  }

  save() {
    if (window.localStorage) {
      window.localStorage.setItem('ShoppingList',
        JSON.stringify({ list: this.list, count: this.count }));
    }
    this.dispatchEvent('update', this.list);
  }

  clear() {
    this.list = [];
    this.save();
  }
}

const service = new ShoppingListServiceSingleton();
export const ShoppingListService = service;
