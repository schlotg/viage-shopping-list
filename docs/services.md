# Viage Shopping List Tutorial

## Services
In Viage, Services are used to manage data or any other global resource. Components should only contain other components and the glu-logic to display data to the user. In contrast, Services should not contain any display logic and handle all aspects of managing the data. This includes notifying components when the data has changed. Services are singletons and do need to be instantiated, you can use them directly by importing them.

### Create a Service
Using the Viage CLI, create a new service by typing the following in a shell in your project's direrctory:

``` viage service shopping-list-service```

This will create a new file at src/services/shopping-list-service.ts. Open this file and it should look like this:

```Javascript
import { Service } from 'viage';

class ShoppingListServiceSingleton extends Service {
  constructor() {
    super();
  }
}
const service = new ShoppingListServiceSingleton ();
export const ShoppingListService = service;
```

Services extend the base class Service, and then export a single instance of the service. To use it in a component you just need to import it:

```Javascript
import { ShoppingListService } from '../service/shopping-list-service';
```

### Defining a Data interface
Let's start by defining an interface for the data. Insert the following lines just above the class declaration:

```Javascript
import { Service } from 'viage';

export interface Item {
  name: string;
  description: string;
  quantity: number;
  purchased: boolean;
  _id?: string; // added by the service
}

class ShoppingListServiceSingleton extends Service {
```

Each Shopping list item is going to have a name, description, quantity, and a purchased flag to indicate if it has been purchased or not. The _id field will be used internally by the service to uniquely track each item. We export the interface so it can be used by components interacting with the data.

### Data Retrieval and Storage
Normally a service would save and retrieve data from a server. Viage doesn't try to duplicate functionality that is already in the DOM. I reccomend using the [fetch API](https://developer.mozilla.org/en-US/docs/Web/API/WindowOrWorkerGlobalScope/fetch). It is an excellant modern interface that supports promises and is available across the 4 big browsers.

Since writing server code is beyond the scope of this tutorial, we will write to the browser's local storage instead. Add the following code to the shopping-list-service:

```Javascript
class ShoppingListServiceSingleton extends Service {

  private list: Item[] = [];
  private count = 1;

  private generateId() {
    this.count += 1;
    return this.count.toString(16);
  }

  constructor() {
    super();
    if (window.localStorage) {
      const json = window.localStorage.getItem('ShoppingList');
      if (json) {
        const obj = JSON.parse(json) || {};
        this.count = obj.count || 0;
        const list = obj.list || [];
        list.forEach((e: Item) => this.addItem(e));
      }
    }
  }

  save() {
    if (window.localStorage) {
      window.localStorage.setItem('ShoppingList',
        JSON.stringify({ list: this.list, count: this.count }));
    }
    this.dispatchEvent('update', this.list);
  }
```

Our service now contains a local list of the data in a variable named *list*. There is also a local variable called *count* that the *generateId()* function uses to generate unique ids. The *contructor()* and the *save()* function contain code that stores and retrieves data from local storage.

Notice that everytime the *save()* function is called, we dispatch an event called *update*, that lets everyone that is listening know that the data has changed. The *dispatchEvent()* function is defined in the *Service* base class along with the ability to add and remove event listeners. More on that later.

### Getting and Setting Data
Next, we will add some getters and setters to make it easy to add, get, and remove data for the components. This is just standard Typescript. Add the following lines to our service:

```Javascript
  private findIndex(id: string): number {
    let i = 0;
    const list = this.list;
    for (; i < list.length; ++i) {
      if (list[i]._id === id) { break; }
    }
    return i;
  }

  addItem(item: Item) {
    item._id = item._id || this.generateId();
    this.list.push(item);
    this.save();
  }

  removeItem(id: string) {
    const i = this.findIndex(id);
    this.list.splice(i, 1);
    this.save();
  }

  getItem(id: string): Item {
    const i = this.findIndex(id);
    return this.list[i];
  }

  forEach(cb: (item: Item, i?: number) => void) {
    this.list.forEach(cb);
  }

  clear() {
    this.list = [];
    this.save();
  }
```

The function *findIndex* is just a helper function to find an element by id so that we can remove and get items. There is also a *forEach()* function that makes it easy to iterate across all of the items in our list.

Whew! That is the biggest file in our project but it actually does a lot and is great foundation for the rest of the app. All we need to do now is bolt up some components to display the data.

The final file should look like this:

```Javascript
import { Service } from 'viage';

export interface Item {
  name: string;
  description: string;
  quantity: number;
  purchased: boolean;
  _id?: string; // added by the service
}

class ShoppingListServiceSingleton extends Service {

  private list: Item[] = [];
  private count = 1;

  private generateId() {
    this.count += 1;
    return this.count.toString(16);
  }

  constructor() {
    super();
    if (window.localStorage) {
      const json = window.localStorage.getItem('ShoppingList');
      if (json) {
        const obj = JSON.parse(json) || {};
        this.count = obj.count || 0;
        const list = obj.list || [];
        list.forEach((e: Item) => this.addItem(e));
      }
    }
  }

  save() {
    if (window.localStorage) {
      window.localStorage.setItem('ShoppingList',
        JSON.stringify({ list: this.list, count: this.count }));
    }
    this.dispatchEvent('update', this.list);
  }

  private findIndex(id: string): number {
    let i = 0;
    const list = this.list;
    for (; i < list.length; ++i) {
      if (list[i]._id === id) { break; }
    }
    return i;
  }

  addItem(item: Item) {
    item._id = item._id || this.generateId();
    this.list.push(item);
    this.save();
  }

  removeItem(id: string) {
    const i = this.findIndex(id);
    this.list.splice(i, 1);
    this.save();
  }

  getItem(id: string): Item {
    const i = this.findIndex(id);
    return this.list[i];
  }

  forEach(cb: (item: Item, i?: number) => void) {
    this.list.forEach(cb);
  }

  clear() {
    this.list = [];
    this.save();
  }
}

const service = new ShoppingListServiceSingleton();
export const ShoppingListService = service;

```

### [Next (Adding a Shopping List Element Component)](shopping-list-element.md)