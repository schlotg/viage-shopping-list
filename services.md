# Viage Shopping List Tutorial

## Services
In Viage Services are used to manage data or any other global resource. Components should only contain other components and the glu-logic to display data to the user. In contrast Services should not contain and display logic and handle all aspects of managing the data that is being displayed including notifying components when data has changed. Services are singletons and do need to be instantiated, you cna use them directly by importing them.

### Create a Service
Using the Viage CLI, create a new service by typing the following in a shell in your project's direrctory:

``` viage service shopping-list-service```

This will create a new file at src/services/shopping-list-service.ts. Open this file and it should look like this:

```Javascript
import { Service } from 'viage';

class ShoppingListServiceSingleton extends Service {
  constructor() {
    super();
  }
}
const service = new ShoppingListServiceSingleton ();
export const ShoppingListService = service;
```

Services extend the base class Service, and then exports a single instance of the service. To use it in a component you just need to import it:

```Javascript
import { ShoppingListService } from '../service/shopping-list-service';
```

### Defining a Data interface
Let's start by defining an interface for the data. Insert the following lines just above the class declaration:

```Javascript
import { Service } from 'viage';

export interface Item {
  name: string;
  description: string;
  quantity: number;
  purchased: boolean;
  _id?: string; // added by the service
}

class ShoppingListServiceSingleton extends Service {
```

Each Shopping list item is going to have a name, description, quantity, and a purchased flag to indicated if purchased or not. The _id field will be used internally by the service to uniquely track each item. We export the interface so it can be used by components interacting with the data.



