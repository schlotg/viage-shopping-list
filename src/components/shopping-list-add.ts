import { Component } from 'viage';
import { ShoppingListService } from '../services/shopping-list-service';
import { getRouter } from 'viage';

export class ShoppingListAdd extends Component {

  params = {id: ''};
  fields = ['quantity', 'name', 'description'];

  constructor(params: any) {
    super('shopping-list-add');
    this.setHTML(`
      <h3>${params && params.id ? 'Edit' : 'Add Item'}</h3>
      <table>
        <tr>
          <th>Quantity</th>
          <th>Name</th>
          <th>Description</th>
        </tr>
        <tr>
          <td><input type="number" attach="quantity" style="width: 60px" /></td>
          <td><input type="text" attach="name" style="width: 100px" /></td>
          <td><input type="text" attach="description" style="width: 300px" /></td>
        </tr>
      </table>
      <button attach="save">Save</button>
      <button attach="back">Back</button>
    `);
    const attachments = this.attachments;
    this.params = params;
    if (params && params.id) {
      this.updateItem();
    }

    attachments.save.addEventListener('click', () => {
      const item: any = {enabled: false};
      this.fields.forEach(k => item[k] = attachments[k].value);
      if (!this.params.id) {
        ShoppingListService.addItem(item);
      } else {
        const data: any = ShoppingListService.getItem(this.params.id);
        this.fields.forEach(k => data[k] = item[k]);
        ShoppingListService.save();
      }
      getRouter('main').go('#home');
    });

    // handle back
    attachments.back.addEventListener('click', () => getRouter('main').back());
  }

  updateItem() {
    const data: any = ShoppingListService.getItem(this.params.id);
    this.fields.forEach(k => this.attachments[k].value = data && data[k]);
  }
}