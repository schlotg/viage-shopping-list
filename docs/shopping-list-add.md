# Viage Shopping List Tutorial

## Creating the Shopping List Add Component
This component will be the view that is used to add an item to the Shopping List. Start by using the CLI to create a *shopping-list-add component*.

``` viage component shopping-list-add ```


Now modify the code so that it looks lke this:

```Javascript
import { Component } from 'viage';
import { ShoppingListService } from '../services/shopping-list-service';
import { States } from './app';

interface ComponentParams {
  id?: string;
}

export class ShoppingListAdd extends Component {

  params: ComponentParams = {id: ''};
  fields = ['quantity', 'name', 'description'];

  constructor() {
    super('shopping-list-add');
  }

  init(params: ComponentParams) {
    this.setHTML(`
      <h3>Add Item</h3>
      <table>
        <tr>
          <th>Quantity</th>
          <th>Name</th>
          <th>Description</th>
        </tr>
        <tr>
          <td><input type="number" attach="quantity" style="width: 60px" /></td>
          <td><input type="text" attach="name" style="width: 100px" /></td>
          <td><input type="text" attach="description" style="width: 300px" /></td>
        </tr>
      </table>
      <button attach="save">Save</button>
      <button attach="back">Back</button>
    `);
    const attachments = this.attachments;

    attachments.save.addEventListener('click', () => {
      const item: any = {purchased: false};
      this.fields.forEach(k => item[k] = attachments[k].value);
      item.quantity = parseInt(item.quantity); // ensure this is a number
      ShoppingListService.addItem(item);
      const homeUrl = this.router.createUrl<void>(States.HOME);
      this.router.go(homeUrl);
    });

    // handle back
    attachments.back.addEventListener('click', () => this.router.back());
    return this;
  }
}
```
### Attach attribute
This is very much like the other components that we have created. The HTML is simply a table of fields that can be edited. This was covered previously but I want to re-iterate. Notice a **attach="save"** and a **attach="back"** attribute in the button HTML. This is the only custom HTML attribute that Viage uses. It has very simple functionality. Putting an attach property in the HTML will instruct the **setHTML()** function to get the element from the DOM, once it is added, and put it in a collection of elements named **attachments**. This is merely a helper to allow the programmer quick and easy access through the **this.attachment** member. As an example, to access the save button merely write something like this:
```Javascript
  const save = this.attachments.save;
```

### Save Button
The *save* button has a click handler added to it and it gathers each input field and constructs an object that is sent to the *ShoppingListService.addItem()* function. Once the call is complete, it calls the router to go to the *home* state. In a app that calls a server to save, you might want to configure the service to return a promise whenever an async operation is performed.

### Back Button
The back button calls the *router.back()* function which tells the router to go back to the previous state or to the default state if no previous state exists. This is emulating the back button functionality on the browser. If fact if you hit the back button on the browser it will accomplish the same thing.

### Installing a Route
The only thing left is to let the router know about this new state. Add the following code into the src/components/app.ts file so that the router states match this:

```Javascript
router.addStates([
      { name: States.HOME, component: ShoppingList,  type: 'DEFAULT' },
      { name: States.ADD, component: ShoppingListAdd,  type: 'NORMAL' },,
    ]);
```

Don't forget to import the ShoppingListAdd component by adding the following to the top of the src/components/app.ts file:

```Javascript
import { ShoppingListAdd } from './shopping-list-add';
```

### Test it out
Finally we can add items to our shopping list and then display them. Press the add button on the app running in the browser and add some items. Your shopping-list-add screen should look like this:

![img2](img2.png)

And the populated list should now look something like this:

![img3](img3.png)

You should be able to Add, Delete, View, Mark as Purchased, and Clear the list. You can even refresh the browser and it will keep the list intact because the contents are saved and retrieved from Localstorage.

### Editing
The only functionality missing is the ability to edit an item.

### [Next (The Shopping List Edit Component)](shopping-list-edit.md)
