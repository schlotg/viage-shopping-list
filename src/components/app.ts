import { ShoppingList } from './shopping-list';
import { ShoppingListAdd } from './shopping-list-add';
import { createRouter, Component } from 'viage';

export enum States {
  HOME = 'home',
  ADD = 'add',
  EDIT = 'edit'
};

export class App extends Component {

  private title = 'Shopping List';

  constructor() {
    super('app');
  }
  init() {
    document.querySelector('title').textContent = this.title;

    this.attach('page', true);
    this.setHTML(`
      <h1 style="text-align: center">${this.title}</h1>
      <div style="width: 100%; min-width: 654px" attach="portal"></div>
    `);

    const router = createRouter('main', this.attachments.portal, 'HASH');
    router.addStates([
      { name: States.HOME, component: ShoppingList,  type: 'DEFAULT' },
      { name: States.ADD, component: ShoppingListAdd,  type: 'NORMAL' },
      { name: States.EDIT, component: ShoppingListAdd,  type: 'NORMAL' },
    ]);
    // start off by going to the state the page was loaded on
    router.start();
    return this;
  }
}
