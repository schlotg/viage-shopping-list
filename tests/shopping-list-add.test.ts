import { ShoppingListAdd } from '../src/components/shopping-list-add';
import { Item, ShoppingListService } from '../src/services/shopping-list-service';
import { createRouter, destroyRouter } from 'viage';
import { item, triggerClickEvent } from './common';

test('ShoppingListAdd should initialize in add mode correctly', () => {
  const add = new ShoppingListAdd();
  const mock = jest.spyOn(add, 'updateItem');
  add.init({ id: '' });
  expect(add.e.innerHTML).toMatchSnapshot();
  expect(add.e.querySelector('h3').textContent).toEqual('Add Item');
  expect(mock).not.toBeCalled();
  mock.mockRestore()
});

test('ShoppingListAdd should initialize in edit mode correctly', () => {
  const add = new ShoppingListAdd();
  const mock = jest.spyOn(add, 'updateItem');
  const mock2 = jest.spyOn(ShoppingListService, 'getItem');
  mock2.mockImplementation(() => item);
  add.init({ id: '5' });
  expect(add.e.innerHTML).toMatchSnapshot();
  expect(add.e.querySelector('h3').textContent).toEqual('Edit');
  expect(mock).toBeCalled();
  mock.mockRestore();
  mock2.mockRestore();
});

test('back event listener should trigger router.back()', () => {
  const add = new ShoppingListAdd();
  const router = createRouter('test', document.createElement('div'), 'LOCATION' );
  const mock = jest.spyOn(router, 'back');
  add.setRouter(router);
  add.init({ id: '' });
  triggerClickEvent(add.e.querySelector('[attach=back]'));
  expect(mock).toBeCalled();
  mock.mockRestore();
  destroyRouter('test');
});

test('Save click handler in add mode should call ShoppingListService.addItem() and call router.go()', () => {
  const add = new ShoppingListAdd();
  const router = createRouter('test', document.createElement('div'), 'LOCATION' );
  const goMock = jest.spyOn(router, 'go');
  const addItemMock = jest.spyOn(ShoppingListService, 'addItem');
  let savedItem: Item;
  addItemMock.mockImplementation((item: Item) => savedItem = item);
  add.setRouter(router);
  add.init({ id: '' });

  // add input data
  for(let k in item) {
    const e = add.e.querySelector(`[attach=${k}]`) as HTMLInputElement;
    if (e) { e.value = item[k]; }
  }

  triggerClickEvent(add.e.querySelector('[attach=save]'));
  expect(addItemMock).toBeCalled();
  expect(goMock).toBeCalled();
  expect(savedItem).toEqual(item);
  goMock.mockRestore();
  addItemMock.mockRestore();
});

test('Save click handler in edit mode should call ShoppingListService.getItem() and then ShoppingListService.save()', () => {
  const add = new ShoppingListAdd();
  const router = createRouter('test', document.createElement('div'), 'LOCATION' );
  const goMock = jest.spyOn(router, 'go');
  ShoppingListService.addItem(item);
  const saveMock = jest.spyOn(ShoppingListService, 'save');
  add.setRouter(router);
  add.init({ id: item._id });

  const modifedItem: any = { name:'Eggs',  description: 'Free Range', quantity: 12 };

  // add input data
  for(let k in modifedItem) {
    const e = add.e.querySelector(`[attach=${k}]`) as HTMLInputElement;
    if (e) { e.value = modifedItem[k]; }
  }

  modifedItem.purchased = false;
  modifedItem._id = item._id;

  triggerClickEvent(add.e.querySelector('[attach=save]'));
  expect(saveMock).toBeCalled();
  expect(goMock).toBeCalled();
  expect(modifedItem).toEqual(ShoppingListService.getItem(modifedItem._id));
  goMock.mockRestore();
  saveMock.mockRestore();
});
