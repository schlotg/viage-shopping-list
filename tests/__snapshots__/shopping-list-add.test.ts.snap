// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ShoppingListAdd should initialize in add mode correctly 1`] = `
"
      <h3>Add Item</h3>
      <table>
        <tbody><tr>
          <th>Quantity</th>
          <th>Name</th>
          <th>Description</th>
        </tr>
        <tr>
          <td><input type=\\"number\\" attach=\\"quantity\\" style=\\"width: 60px\\"></td>
          <td><input type=\\"text\\" attach=\\"name\\" style=\\"width: 100px\\"></td>
          <td><input type=\\"text\\" attach=\\"description\\" style=\\"width: 300px\\"></td>
        </tr>
      </tbody></table>
      <button attach=\\"save\\">Save</button>
      <button attach=\\"back\\">Back</button>
    "
`;

exports[`ShoppingListAdd should initialize in edit mode correctly 1`] = `
"
      <h3>Edit</h3>
      <table>
        <tbody><tr>
          <th>Quantity</th>
          <th>Name</th>
          <th>Description</th>
        </tr>
        <tr>
          <td><input type=\\"number\\" attach=\\"quantity\\" style=\\"width: 60px\\"></td>
          <td><input type=\\"text\\" attach=\\"name\\" style=\\"width: 100px\\"></td>
          <td><input type=\\"text\\" attach=\\"description\\" style=\\"width: 300px\\"></td>
        </tr>
      </tbody></table>
      <button attach=\\"save\\">Save</button>
      <button attach=\\"back\\">Back</button>
    "
`;
